snapshot.go

```diff
@@ -146,6 +146,7 @@ func (r *Raft) takeSnapshot() (string, error) {
 	// We have to use the future here to safely get this information since
 	// it is owned by the main thread.
 	configReq := &configurationsFuture{}
+	configReq.ShutdownCh = r.shutdownCh
```

Comment:
It may be nice to document when `ShutdownCh` needs to be set and what makes snapshotting special.  I assume it's because snapshotting runs in an internal gorotouine and can run regardless of leadership.  Other futures will be responded to immediately when not leader, or when raft loses leadership as a result of shutting down?